### DISCUSSION TITLE
动态切换全局配置的方法是不是和Plugin-Model有冲突？

### DISCUSSION THREAD

**Discussion by DreamParadise** [umijs/umi#12287] [Q&A] [answered]
# 相关依赖
```json
{
   "dependencies": {
        "antd": "^5.15.1",
        "antd-style": "^3.6.1",
        "react": "^18.2.0",
        "react-dom": "^18.2.0",
        "react-helmet-async": "^1.3.0",
    },
    "devDependencies": {
        "@ant-design/pro-cli": "^3.3.0",
        "@types/react": "^18.2.48",
        "@types/react-dom": "^18.2.18",
        "@types/react-helmet": "^6.1.11",
        "@umijs/fabric": "^2.14.1",
        "@umijs/lint": "^4.1.1",
        "@umijs/max": "^4.1.9",
    }
}
```
# 调用方法
```tsx
  const changeTheme = () => {
    const { themeMode, antdToken } = getThemeToken();
    const targetThemeMode = themeMode === 'dark' ? 'light' : 'dark';
    // 将变量写入localStorage
    setThemeMode(targetThemeMode);
    const algorithm = targetThemeMode === 'light' ? [defaultAlgorithm] : [darkAlgorithm];
    const antdTheme: ThemeConfig = { algorithm };
    if (antdToken) {
      antdTheme.token = antdToken;
    }
    setAntdConfig({ theme: antdTheme });
```
# 报错信息
![image](https://github.com/umijs/umi/assets/3191003/23499254-468e-43b5-816f-50c72a97ed3c)

**Comment by fz6m** [comment]
antd 主题更换的逻辑应该可以正常运行，就是不知道 model 数据流里的东西会不会受影响。

这个 model 数据流是上古的方案一直没维护，只是为 umi 3 升级兼容而保留的，和现在的前端环境格格不入了；现代全局数据流更多人都使用了 [zustand](https://github.com/pmndrs/zustand) 、[jotai](https://github.com/pmndrs/jotai) 、[valtio](https://umijs.org/docs/max/valtio) ，或者选择 [hox](https://github.com/umijs/hox) 这种类似 model 的方案，不推荐再使用内置的这个 model 数据流功能了。

如有兴趣，可以探究下为什么会发生这种问题，因为我看出现报错的地方在 model 里的一个 `setState` （展开错误堆栈可以找到），也许搞懂原因给 model 的逻辑修复兼容下就可以了。

**Reply by DreamParadise** [reply]
1. 是的，样式是成功切换了，但是有些组件的功能受影响了。我有个下拉框的选项是从model中获取的，切换完主题后这个组件下拉选项没了。
2. 他不止和我们自己在`src/models`目录下自定义的model有冲突，和`export async function getInitialState()`也有冲突，这个全局初始状态好多地方都用到了，这不好改吧，应该还是得解决这个问题。

**Reply by DreamParadise** [reply]
我的项目是使用ant design pro创建的，initial state在布局组件和权限管理组件都用到了，如果按你说的使用其它的全局数据流方案的话应该要怎么改造呢？比如这些`export const layout: RunTimeLayoutConfig = ({ initialState, loading, refresh }) => {}`、`export default function access(initialState: any) {}`

**Reply by fz6m** [reply]
`getInitialState` 默认就是一个 model ，为了代替这个行为，你可以在 render 前获取数据，然后放到 valtio 里（valtio 是一种和 react 无关的状态管理库），之后在任意地方使用即可：

```ts
// src/app.ts

export const render = async (oldRender: any) => {
  // call api
  oldRender()
}
```

或者加一个 context ：

```ts
// src/app.tsx

export const rootContainer = (container: any) => {
  return <CustomContext>{container}</CustomContext>
}
```

这样你在 `CustomContext` 里可以在 `useEffect` 里获取数据，获取完数据再渲染内容，数据存到 zustand / jotai 里面供全局使用即可（使用这两个库时注意添加对应库的 provider context 到最外层）。

通过使用 1 种全局状态管理库来替换 model 即可。

如果想保持以前 model 的风格，在 model 里写 hooks ，就用 hox 吧，hox 类似于一个局部性的 model ，具体可自行查看文档。

**Reply by DreamParadise** [reply]
好的，多谢，我尝试下。

**Reply by DreamParadise** [reply]
@fz6m 你好，我把model和initialState插件都关闭之后，路由的权限该怎么配置呢，我看access插件好像也是用了model吧？还有`app.tsx`中`layout`的运行时配置的入参是不是也没用了，我看原来的入参是从initialState拿的数据？

**Reply by DreamParadise** [reply]
@fz6m 
1. 使用jotai的话umijs的运行时配置里面是不是取不到值？因为运行时配置的属性值都不是React Hook，比如layout本来是直接在方法的参数把initialState传进来的话换成jotai之后我怎么从jotai拿到这个值。
2. 路由权限怎么配置呢？原来框架是使用access插件的，但是这个插件依赖model，我现在把model关闭了

**Reply by fz6m** [reply]
你要用 access 权限的话，就没法关 model ，因为 access 依赖了 model 和 initialState 。

jotai 只能在 react hooks 里用，得想其他方法去解决问题。

我又重新看了下 model 报错， antd config provider 的更新导致整个组件树重新渲染了，走到 model 的那一层 context 的时候，组件重新挂载，导致在重渲染的过程中执行了 `setState` （因为 model 的 context 里在第一次挂载的时候，急切的需要执行一次获取初始值的操作，而这个操作在执行的时候，正好处于 antd 重渲染的流程中），react 不允许这种行为。

调了很久不知道怎么解决，我目测还是要从 model 的源码来调整，想办法解决这个问题，但 model 这个功能太上古而且太 hack 了，如果你有兴趣的话可以继续研究下这个地方应该怎么避免。

试试在 `app.tsx` 的 `export const innerProvider` 这个层级自己添加一个 antd ConfigProvider 来解决这个问题吧，就不用 setAntdConfig 这个默认提供的了。

**Reply by DreamParadise** [reply]
@fz6m 辛苦大佬了！
我现在计划现在用`jotai`状态库来替换model，就是有个问题不知道`access`权限怎么解决。我看了`plugin-layout`的代码是写死通过`access`插件的`useAccessMarkedRoutes`来过滤路由权限的，有没有什么办法我自己实现一个`useAccessMarkedRoutes`替换进去，这样我就可以彻底不用model了。
前面提到的运行时配置中使用react hooks方法IDE报错提示的，虽然有提示不过好像功能是正常的，能正常取到值😂

**Reply by fz6m** [reply, marked as answer]
先这样吧，在项目根目录创建一个 `plugin.ts` 项目级插件：

```ts
// plugin.ts

import { IApi } from '@umijs/max';

export default (api: IApi) => {

  api.onGenerateFiles({
    stage: Infinity,
    fn() {
      api.writeTmpFile({
        path: 'plugin-access/runtime.tsx',
        noPluginDir: true,
        content: `
    import React from 'react';
    import { AccessContext } from './context';
    import useInitialAccess from '@/useInitialAccess';
    
    function Provider(props) {
      const access = useInitialAccess();
    
      return (
        <AccessContext.Provider value={access}>
          { props.children }
        </AccessContext.Provider>
      );
    }
    
    export function accessProvider(container) {
      return <Provider>{ container }</Provider>;
    }
    `
      })
    }
  })

};

```

关闭 `model` 和 `initialStage` ，只打开 `access` ，之后在 `src/useInitialAccess.ts` 这个 hook 里取一下你的 jotai 值就行，里面的权限 kv 找个比较早的时机设定好，或者你再想办法改造下。

这个方法比较 hack ，我还没想好怎么解决 model 和 antd 重渲染的冲突，最好还是改 model 里面运作的代码解决。

**Comment by fz6m** [comment]
这个问题知道为什么了，详见 #12406 。

在 PR 没修复并发布新版本之前，只要自己给 antd 的配置加个空 `theme: {}` 对象即可：

```ts
// .umirc.ts

  antd: { configProvider: { theme: {} } }
```

这里为空对象不会影响什么，antd 源码里如果不配置 theme ，默认也会用空对象。

之前发过的那个项目级插件写文件的解法就不要用了，太 hack 了。

**Reply by DreamParadise** [reply]
我已经按之前回复的把整个model都移除了😂

**Reply by fz6m** [reply]
这个问题已经修复了，并且在最新的版本发布了，如果需要 model 的话，可以继续使用，感谢你的反馈。
